src/main/ruby/core/process_mirror.rb

```diff
@@ -173,18 +173,41 @@ def initialize(env_or_cmd, *args)
           end
 
           @options = {}
+          @files_for_child = []
 
           if options
             options.each do |key, value|
               case key
               when ::IO, ::Fixnum, :in, :out, :err
                 from = convert_io_fd key
-                to = convert_to_fd value, from
+                to = convert_to_fd value, from, @files_for_child
                 redirect @options, from, to
               when ::Array
-                from = convert_io_fd key.first
-                to = convert_to_fd value, from
-                key.each { |k| redirect @options, convert_io_fd(k), to }
+
+                # When redirecting multiple fds to one file, as in
+                #
+                # [:out, :err] => "/some/file"
+                #
+                # We want this to be equivalent to
+                #
+                # :hidden_fd => "/some/file", :out => :hidden_fd, :err => :hidden_fd
+                #
+                # opening the destination once, and only then
+                # redirecting everything to that destination.
+                #
+                # Therefore we open all files in the parent process,
+                # and only specify redirection for the child process.
+                #
+                # We don't use open actions for this as we want to
+                # guarantee the order and need to work around a bug in
+                # old versions of glibc which do not correctly
+                # duplicate the string containing the file path.
+
+                key.each do |k|
+                  from = convert_io_fd(k)
+                  to = convert_to_fd value, from, @files_for_child unless to
```

Comment:
This is a bit magic as it relies on `to` being declared in the scope above.
How about:
```ruby
fds = key.map { |k| convert_io_fd(k) }
to = convert_to_fd value, fds.first, @files_for_child
fds.each { |fd| redirect @options, fd, to }
```